Support sending and receiving file descriptors over Unix domain sockets

mattrm456 opened the pull request:
The Unixes and Unix-likes support sending an open file descriptor between two processes using Unix domain sockets. The receiving process receives a reference to an open kernel object as if it called `dup()` on the original file descriptor. Users have expressed a desire to send file descriptors over NTS and NTC sockets. This PR adds support to NTS socket to send and receive file descriptors over Unix domain sockets. Support for similar functionality in NTC will be proposed in a subsequent PR.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
+
+ntsa::Error SendControl::encode(msghdr* msg, const ntsa::SendOptions& options)
+{
+    bsl::memset(d_arena.buffer(), 0, k_SEND_CONTROL_BUFFER_SIZE);
```
I think it makes sence to put this `memset` line after the `if (options.foreignHandle().isNull())` block. What's the point in zeroing the memory if we have nothing to encode?

mattrm456 replied:
Done.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
+                    context->setForeignHandle(foreignHandle);
+                }
+                else {
+                    ::close(foreignHandle);
```
Maybe we should have here WARN of DEBUG level print? Otherwise there is no way at all to know that a foreigh process wanted to send us a file descriptor while we were not interested in that.

mattrm456 replied:
Done.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
+            else if (hdr->cmsg_type == TimestampUtil::e_SCM_TIMESTAMPING) {
                 TimestampUtil::ScmTimestamping ts;
 
                 if (NTSCFG_UNLIKELY(sizeof(ts) !=
```
You added a WARN level logging for the case above when we ignore a file descriptor control message due to unexpected payload size. Here we ignore a timestamp due to the following  reason, but there is no logging. Is it ok to fix this discripancy?

mattrm456 replied:
Done.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -1198,17 +1361,11 @@ ntsa::Error SocketUtil::send(ntsa::SendContext*       context,
 
+    context->setBytesSendable(size);
+
     ssize_t sendmsgResult =
```
sendmsgResult as a counter of bytes does not include control message bytes, right?

mattrm456 replied:
No, just real data.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -2361,46 +2525,39 @@ ntsa::Error SocketUtil::receive(ntsa::ReceiveContext*       context,
+
+    bsl::size_t capacity =
+        ntsa::MutableBuffer::totalSize(bufferArray, bufferCount);
+    if (capacity == 0) {
```
Here and and many other places this validation is now not the first thing `receive` function does. I think it is better to perform this check immediately and avoid unnecessray work done in a failure case.

mattrm456 replied:
That is true, but I very much prefer the structure of the functions as they are written:
```
<common-setup>
<data-representation-specific-setup>
recvmsg
<common-cleanup>
```
We have a lot of overloads here and the functions as-written help me visually orient and detect that <common-setup> is correctly performed in each function. Calling `ntsu::SocketUtil::receive` with no capacity into which the copy the data is essentially an invalid call that I'm not particularly motivated by, given the previously described concern.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -2424,11 +2581,9 @@ ntsa::Error SocketUtil::receive(ntsa::ReceiveContext*       context,
-    if (wantTimestamp) {
-        extractTimestampsFromCtrlBlock(context, msg);
+    if (wantMetaData) {
+        control.decode(context, msg, options);
```
I have a general concern and I don't know what to suggest as a solution. `receive` functions do not return any actual error if there is something wrong during decoding. And while it may be ok for RX timestamps, I think it's not ok for file descriptors transfer. And unfortunately there is no way to tell the user something like "in general data reception was performed correctly and you can use this data, but there also was an attempt to receive a control message and it failed". Maybe `receive` functions should return a tuple of `ntsa::Error`s?

mattrm456 replied:
This is a good observation. However, `ReceiveControl::decode` doesn't actually have a failure mode, as currently implemented (and it has the same problem: we can "fail" to decode a file descriptor but successfully decode a timestamp.) An API that returns like an array of ntsa::Error sounds very unpleasant. For now, I think we should live with the (new) BSLS_LOG_WARN logging alert users that while reception of the data was successful, some reception of the meta-data was not.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.cpp:
```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
+        // file descriptors, timestamps, etc.) buffered by the operating system
+        // for a socket.
+        k_RECEIVE_CONTROL_PAYLOAD_SIZE = static_cast<int>(
+            NTSU_SOCKETUTIL_MAX_HANDLES_PER_INCOMING_CONTROLMSG * sizeof(int))
```
`sizeof(int)`  is everywhere to mean amount of memory required to represent transferred file handle. I suggest to either use an explicit constant instead of this `sizeof` or to define a new typedef and use it instead of just `int`.

mattrm456 replied:
No Unix that I know of defines a typedef for the type of a file descriptor, it's just an `int`. However, we typedef `ntsa::Handle` to `int` on Unixes. I have revised most places of the code to use `ntsa::Handle` instead of `int` and either use `sizeof <variable>` or `sizeof(ntsa::Handle)`.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.t.cpp:
```diff
@@ -616,6 +709,221 @@ void testStreamSocketTransmissionBlob(ntsa::Transport::Value transport,
+        bdlbb::BlobUtil::append(&clientBlob, DATA, sizeof DATA - 1);
+
+        bdlbb::Blob serverBlob(&blobBufferFactory, allocator);
+        serverBlob.setLength(sizeof DATA - 1);
```
It seems to be not needed. The next line sets blob's length to 0. I see the same pattern in other tests where `bdlbb::Blob` is used.

mattrm456 replied:
This is the idiom for `bdlbb::Blob` to `reserve(<size>)`.

smtrfnv commented on groups/nts/ntsu/ntsu_socketutil.t.cpp:
```diff
@@ -616,6 +709,221 @@ void testStreamSocketTransmissionBlob(ntsa::Transport::Value transport,
+                // passed using socket control messages that is not "picked up"
+                // immediately when it is available is not subsequently
+                // retrievable with a subsequent call to 'recvmsg'.
+#if 0
```
It seems like you forgot to remove this code. And the same is applicable to another (datagram based) test

mattrm456 replied:
Yes this is dead code. I've deleted this preprocessed-out block in these two places.

smtrfnv reviewed (commented):
Point to discuss: technically UDS datagram sockets support RX timetamping. Maybe we can have a test to check that RX timestamp and a file handle can be received via one control message.

mattrm456 commented:
I have addressed all your review comments except for the one about partially setting up `recvmsg` before checking the user supplies valid capacity, and your comment about a UDS test to ensure timestamps and file descriptors can be passed together in the same control message.

smtrfnv reviewed (approved):
LGTM
